src/style-spec/validate_mapbox_api_supported.js

```diff
@@ -0,0 +1,172 @@
+    if (!isValid(source.url, sourceUrlPattern)) {
+        errors.push(new ValidationError(`sources[${i}]`, source.url, 'Style must reference sources hosted by Mapbox'));
+    }
+
+    return errors;
+}
+
+function getSourcesErrors(sources: Object): Array<?ValidationError> {
+    const errors = [];
+    let count = 0;
+
+    Object.keys(sources).forEach((s: string, i: number) => {
+        const sourceErrors = getSourceErrors(sources[s], i);
+
+        // If source has errors, skip counting
+        if (!sourceErrors.length) {
+            count = count + getSourceCount(sources[s]);
+        }
+
+        errors.push(...sourceErrors);
+    });
+
+    if (count > MAX_SOURCES_IN_STYLE) {
+        errors.push(new ValidationError('sources', null, `Styles must contain ${MAX_SOURCES_IN_STYLE} or fewer sources`));
+    }
+
+    return errors;
+}
+
+function getRootErrors(style: Object, specKeys: Array<any>): Array<?ValidationError> {
+    const errors = [];
+
+    /*
+     * The following keys are optional but fully managed by the Mapbox Styles
+     * API. Values on stylesheet on POST or PATCH will be ignored: "owner",
+     * "id", "cacheControl", "draft", "created", "modified"
+     *
+     * The following keys are optional. The Mapbox Styles API respects value on
+     * stylesheet on PATCH, but ignores the value on POST: "visibility"
+     */
+    const optionalRootProperties = [
+        'owner',
+        'id',
+        'cacheControl',
+        'draft',
+        'created',
+        'modified',
+        'visibility'
+    ];
+
+    const allowedKeyErrors = getAllowedKeyErrors(style, [...specKeys, ...optionalRootProperties]);
+    errors.push(...allowedKeyErrors);
+
+    if (style.version > SUPPORTED_SPEC_VERSION || style.version < SUPPORTED_SPEC_VERSION) {
+        errors.push(new ValidationError('version', style.version, `style version must be ${SUPPORTED_SPEC_VERSION}`));
+    }
+
+    /*
+     * "glyphs" is optional. If present, valid examples:
+     * mapbox://fonts/penny/{fontstack}/{range}.pbf
+     * mapbox://fonts/mapbox/{fontstack}/{range}.pbf
+     */
+    const glyphUrlPattern = /^mapbox:\/\/fonts\/([^/]*)\/{fontstack}\/{range}.pbf$/;
+    if (!isValid(style.glyphs, glyphUrlPattern)) {
+        errors.push(new ValidationError('glyphs', style.glyphs, 'Styles must reference glyphs hosted by Mapbox'));
+    }
+
+    /*
+     * "sprite" is optional. If present, valid examples:
+     * mapbox://sprites/penny/abcd1234
+     * mapbox://sprites/mapbox/abcd1234/draft
+     * mapbox://sprites/cyrus/abcd1234/abcd1234
+     */
+    const spriteUrlPattern = /^mapbox:\/\/sprites\/([^/]*)\/([^/]*)\/?([^/]*)?$/;
+    if (!isValid(style.sprite, spriteUrlPattern)) {
+        errors.push(new ValidationError('sprite', style.sprite, 'Styles must reference sprites hosted by Mapbox'));
+    }
+
+    /*
+     * "visibility" is optional. If present, valid examples:
+     * "private"
+     * "public"
+     */
+    const visibilityPattern = /^(public|private)/;
```

Comment:
I think this needs a `$` at the end to make sure strings like `'private-boop'` are not admitted:

```suggestion
    const visibilityPattern = /^(public|private)$/;
```